packages/asset/contracts/AssetReveal.sol

```diff
@@ -0,0 +1,266 @@
+        );
+    }
+
+    /// @notice Burn multiple assets to be able to reveal them later
+    /// @dev Can be used to burn multiple copies of the same token id, each copy will be revealed separately
+    /// @param tokenIds the tokenIds of the assets to burn
+    /// @param amounts the amounts of the assets to burn
+    function revealBatchBurn(
+        uint256[] calldata tokenIds,
+        uint256[] calldata amounts
+    ) external {
+        require(tokenIds.length == amounts.length, "Invalid input");
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            revealBurn(tokenIds[i], amounts[i]);
+        }
+    }
+
+    /// @notice Reveal assets to view their abilities and enhancements
+    /// @dev Can be used to reveal multiple copies of the same token id
+    /// @param signature Signature created on the TSB backend containing REVEAL_TYPEHASH and associated data, must be signed by authorized signer
+    /// @param creator The original creator of the assets
+    /// @param prevTokenId The tokenId of the unrevealed asset
+    /// @param amount The amount of assets to reveal (must be equal to the length of revealHashes)
+    /// @param metadataHashes The array of hashes for asset metadata
+    /// @param recipient The recipient of the revealed assets
+    function revealMint(
+        bytes memory signature,
+        address creator,
+        uint256 prevTokenId,
+        uint256 amount,
+        string[] memory metadataHashes,
+        address recipient
+    ) public {
+        require(
+            authValidator.verify(
+                signature,
+                _hashReveal(creator, prevTokenId, amount, metadataHashes)
+            ),
+            "Invalid signature"
+        );
+
+        IAsset.AssetData memory data = prevTokenId.getData();
+        require(!data.revealed, "Asset: already revealed");
+        require(
+            data.creator == creator,
+            "Asset: creator does not match prevTokenId"
+        );
+        require(amount == metadataHashes.length, "Invalid amount");
+
+        uint256[] memory tokenIdArray = new uint256[](amount);
+        uint256[] memory tokenCountArray = new uint256[](amount);
+        uint256 uniqueTokenCount = 0;
+
+        // for each asset, set the data
+        for (uint256 i = 0; i < amount; i++) {
+            uint256 tokenId;
+            if (
+                assetContract.getTokenIdByMetadataHash(metadataHashes[i]) != 0
+            ) {
+                tokenId = assetContract.getTokenIdByMetadataHash(
+                    metadataHashes[i]
+                );
+            } else {
+                uint16 revealNonce = ++revealNonces[data.creator][prevTokenId];
+
+                tokenId = TokenIdUtils.generateTokenId(
+                    data.creator,
+                    data.tier,
+                    data.creatorNonce,
+                    revealNonce
+                );
+
+                assetContract.setMetadataHashUsed(tokenId, metadataHashes[i]);
```

Comment:
i don't think the reveal contract should take care of saving the hash,
it should be done in the asset mint & mintbatch maybe

Reply:
Similar to what you've said before and again 100% agree. I have changed it in one of the recent commits.